File: plugins/ui/src/js/src/events/Notification.ts
import { ToastQueue } from '@deephaven/components';
import Log from '@deephaven/log';

const log = Log.module('Notification');

export const NOTIFICATION_EVENT = 'notification.event';

export type NotificationParams = {
  title: string;
  description?: string;
  icon?: string;
  tag?: string;
  silent?: boolean;
  onClick?: () => void;
  onClose?: () => void;
};

/**
 * Show the notification message as a toast. Used as a fallback when the
 * Notifications API is unavailable or permission has not been granted.
 *
 * @param params The notification event parameters
 */
function showToastFallback(params: NotificationParams): void {
  const { title, description, onClick } = params;
  const message = description != null ? `${title}: ${description}` : title;
  ToastQueue.info(
    message,
    onClick != null ? { actionLabel: 'View', onAction: onClick } : undefined
  );
}

/**
 * Handle a notification event by displaying a system notification using the
 * browser's Notifications API.
 *
 * If notifications are not supported or the user has denied permission, the
 * message is shown as a toast instead. If permission has not yet been requested,
 * it will be requested before displaying the notification.
 *
 * @param params The notification event parameters
 */
export async function showNotification(
  params: NotificationParams
): Promise<void> {
  const { title, description, icon, tag, silent, onClick, onClose } = params;

  if (typeof Notification === 'undefined') {
    log.warn('Notifications are not supported, falling back to a toast');
    showToastFallback(params);
    return;
  }

  let { permission } = Notification;
  if (permission === 'default') {
    try {
      permission = await Notification.requestPermission();
    } catch (e) {
      log.warn('Error requesting notification permission', e);
      showToastFallback(params);
      return;
    }
  }

  if (permission !== 'granted') {
    log.debug('Notification permission not granted, falling back to a toast');
    showToastFallback(params);
    return;
  }

  const notification = new Notification(title, {
    body: description,
    icon,
    tag,
    silent,
  });

  if (onClick != null) {
    notification.onclick = () => {
      onClick();
    };
  }
  if (onClose != null) {
    notification.onclose = () => {
      onClose();
    };
  }
}

export default showNotification;

File: plugins/ui/src/deephaven/ui/components/notification.py
from __future__ import annotations

from ..hooks import use_send_event

from typing import Callable
from .._internal.utils import dict_to_react_props
from .._internal.EventContext import NoContextException

_NOTIFICATION_EVENT = "notification.event"


class NotificationException(NoContextException):
    pass


def notification(
    title: str,
    *,
    description: str | None = None,
    icon: str | None = None,
    tag: str | None = None,
    silent: bool | None = None,
    on_click: Callable[[], None] | None = None,
    on_close: Callable[[], None] | None = None,
) -> None:
    """
    Displays a system notification to the user using the browser's Notifications API.

    Notifications appear outside the browser window, at the operating system level, so
    they can reach the user even when the Deephaven tab is not focused. The browser must
    be served over a secure context (HTTPS or localhost) and the user must grant
    permission to display notifications. If permission is denied or notifications are not
    supported, the message is shown as a toast instead.

    Args:
        title: The title to display in the notification.
        description: The body text to display below the title.
        icon: The URL of an image to display as the notification's icon.
        tag: An identifying tag for the notification. Notifications with the same tag
            replace each other instead of stacking, which is useful for updating an
            existing notification.
        silent: Whether the notification should be silent (no sounds or vibrations),
            regardless of the device settings.
        on_click: Handler that is called when the user clicks the notification.
        on_close: Handler that is called when the notification is closed, either by the
            user or after a timeout.

    Returns:
        None
    """
    params = dict_to_react_props(locals())
    try:
        send_event = use_send_event()
    except NoContextException as e:
        raise NotificationException(
            "Notifications must be triggered from the render thread. Use the hook `use_render_queue` to queue a function on the render thread."
        ) from e
    send_event(_NOTIFICATION_EVENT, params)
